File: Sort/75sortColors.py
#三分法

class Solution(object):
    def sortColors(self, nums):
        """
        :type nums: List[int]
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        zero, one, two = -1, 0, len(nums)
        
        while one < two:
            if nums[one] == 0:
                zero += 1
                nums[zero], nums[one] = nums[one], nums[zero]
                one += 1
            elif nums[one] == 2:
                two -= 1
                nums[two],nums[one] = nums[one],nums[two]
            else:
                one += 1

File: DoublePointer/75SortColors.cpp
class Solution {
public:
    void sortColors(vector<int>& nums) {
        if(nums.size()==0) return;
        int l = 0, r = nums.size()-1, index = 0;
        while(index<=r){
            if(nums[index]==0){
                swap(nums[index], nums[l]);
                l++;
                index++;
            }else if(nums[index]==1) index++;
            else {
                swap(nums[index], nums[r]);
                r--;
            }
        }
    }
};

class Solution {
public:
    void sortColors(vector<int>& nums) {
        int l = 0, r = nums.size()-1;
        for(int i=0; i<=r; i++){
            while(nums[i]==2 && i<r) swap(nums[i], nums[r--]);
            while(nums[i]==0 && i>l) swap(nums[i], nums[l++]);
        }
    }
};